components/ws-manager-api/go/crd/v1/workspace_types.go

```diff
@@ -0,0 +1,226 @@
+
+	// +kubebuilder:validation:Required
+	WorkspaceLocation string `json:"workspaceLocation"`
+
+	Git *GitSpec `json:"git,omitempty"`
+
+	Timeout TimeoutSpec `json:"timeout"`
+
+	Admission AdmissionSpec `json:"admission"`
+
+	Ports []PortSpec `json:"ports,omitempty"`
+}
+
+type Ownership struct {
+	// +kubebuilder:validation:Required
+	Owner string `json:"owner"`
+	// +kubebuilder:validation:Required
+	WorkspaceID string `json:"workspaceID"`
+	Team        string `json:"team,omitempty"`
+	Tenant      string `json:"tenant,omitempty"`
+}
+
+type WorkspaceType string
+
+const (
+	WorkspaceTypeRegular    WorkspaceType = "Regular"
+	WorkspaceTypePrebuild   WorkspaceType = "Prebuild"
+	WorkspaceTypeImageBuild WorkspaceType = "ImageBuild"
+)
+
+type WorkspaceImages struct {
+	Workspace WorkspaceImage `json:"workspace"`
+	IDE       IDEImages      `json:"ide"`
+}
+
+type WorkspaceImage struct {
+	Ref *string `json:"ref,omitempty"`
+}
+
+type IDEImages struct {
+	Web        string `json:"web"`
+	Desktop    string `json:"desktop"`
+	Supervisor string `json:"supervisor"`
+}
+
+type GitSpec struct {
+	Username string `json:"username"`
+	Email    string `json:"email"`
+}
+
+type TimeoutSpec struct {
+	Time *metav1.Duration `json:"time,omitempty"`
+}
+
+type AdmissionSpec struct {
+	Level AdmissionLevel `json:"level"`
+}
+
+type AdmissionLevel string
+
+const (
+	AdmissionLevelOwner    AdmissionLevel = "owner"
+	AdmissionLevelEveryone AdmissionLevel = "everyone"
+)
+
+type PortSpec struct {
+	// +kubebuilder:validation:Required
+	Port uint32 `json:"port"`
+
+	// +kubebuilder:validation:Required
+	Visibility AdmissionLevel `json:"visibility"`
+}
+
+// WorkspaceStatus defines the observed state of Workspace
+type WorkspaceStatus struct {
+	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	PodStarts  int    `json:"podStarts"`
+	Headless   bool   `json:"headless"`
+	URL        string `json:"url,omitempty"`
+	OwnerToken string `json:"ownerToken,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	Phase WorkspacePhase `json:"phase,omitempty"`
+	// +kubebuilder:validation:Optional
+	Conditions WorkspaceConditions      `json:"conditions,omitempty"`
+	Results    *WorkspaceResults        `json:"results,omitempty"`
+	Disposal   *WorkspaceDisposalStatus `json:"disposal,omitempty"`
+
+	Runtime *WorkspaceRuntimeStatus `json:"runtime,omitempty"`
+}
+
+type WorkspacePhase string
+
+const (
+	WorkspacePhaseUnknown      WorkspacePhase = "unknown"
+	WorkspacePhasePending      WorkspacePhase = "pending"
+	WorkspacePhaseImageBuild   WorkspacePhase = "imagebuild"
+	WorkspacePhaseCreating     WorkspacePhase = "creating"
+	WorkspacePhaseInitializing WorkspacePhase = "initializing"
+	WorkspacePhaseRunning      WorkspacePhase = "running"
+	WorkspacePhaseStopping     WorkspacePhase = "stopping"
+	WorkspacePhaseStopped      WorkspacePhase = "stopped"
+)
+
+type WorkspaceConditions struct {
```

Comment:
I think we need to talk about this. We should follow the same behavior than k8s proposes, an array, like we can see here https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1beta1/gateway_types.go#L497 (or any k8s objecT)

Reply:
That makes sense - getting closer to common practice is the right way.
What's the convention for making consumers of the API aware of the expected conditions?

Reply:
We need to define that and use `metav1.Condition` `Type` (to define) and use the `Status` to signal in which part of the lifecycle we are and the status, like, WaitingForSupervisor=true or Ring0=false (to signal something wrong with the pod)

I think this is what we need to focus as the first step.